File: DesignPatterns/Creational/Builder/go/Student.go
package main

import "fmt"

type Student struct {
	name   string
	age    int
	height float64
	weight float64
}

type StudentBuilder interface {
	SetName(string) StudentBuilder
	SetAge(int) StudentBuilder
	SetHeight(float64) StudentBuilder
	SetWeight(float64) StudentBuilder
	Build() *Student
}

type studentBuilder struct {
	student *Student
}

func NewStudentBuilder() StudentBuilder {
	student := Student{}
	return &studentBuilder{student: &student}
}

func (st *studentBuilder) SetName(name string) StudentBuilder {
	st.student.name = name
	return st
}
func (st *studentBuilder) SetAge(age int) StudentBuilder {
	st.student.age = age
	return st
}
func (st *studentBuilder) SetHeight(height float64) StudentBuilder {
	st.student.height = height
	return st
}
func (st *studentBuilder) SetWeight(weight float64) StudentBuilder {
	st.student.weight = weight
	return st
}

func (st *studentBuilder) Build() *Student {
	if st.student.age < 16 {
		fmt.Println("Age must be >= 16")
		return nil
	}
	return st.student
}

File: DesignPatterns/Creational/Builder/Student.java
package Creational.Builder;

import java.security.InvalidParameterException;

public class Student {
    private String name;
    private int age;
    private float height;
    private float weight;

    public String get_name(){
        return this.name;
    }

    public int get_age(){
        return this.age;
    }
    public float get_height(){
        return this.height;
    }
    public float get_weight(){
        return this.weight;
    }


    private Student() {}

    public static StudentBuilder getBuilder() {
        return new StudentBuilder();
    }

    public static class StudentBuilder {
        private String name;
        private int age;
        private float height;
        private float weight;


        public StudentBuilder set_name(String name) {
            this.name = name;
            return this;
        }
        public StudentBuilder set_age(int age) {
            this.age = age;
            return this;
        }
        public StudentBuilder set_height(float height) {
            this.height = height;
            return this;
        }
        public StudentBuilder set_weight(float weight) {
            this.weight = weight;
            return this;
        }
        public Student build() throws InvalidParameterException {
            if(this.age < 16) {
                throw new InvalidParameterException("age should be >= 16");
            }
            Student student = new Student();
            student.name = this.name;
            student.age = this.age;
            student.height = this.height;
            student.weight = this.weight;

            return student;
        }
    }

}